feat(browser): new SPA functionality (#15942)
newrelic/docs-website: src/content/docs/browser/single-page-app-monitoring/get-started/spa-v2.mdx

---

> title: Browser SPA monitoring 2.0

Author: ```suggestion
title: Soft navigation support for browser SPA monitoring
```

Author: @cwli24  What do you think of this title?

Reviewer: We technically already support "soft navigations" with the current spa functionality.

tags:
  - Browser
  - Single page app monitoring
translate:
  - jp
metaDescription: The new browser SPA monitoring capability is a streamlined way to monitor your users' experience and app performance.
freshnessValidatedDate: 02-01-2024
---

> We're excited to announce an overhaul of our Pro+SPA agent, built to address numerous pain points voiced by our customers. This update tackles critical limitations of the current agent, paving the way for a smoother, more optimized, and future-proof SPA monitoring experience.

Reviewer: Not sure if this distinction should be made, but we're not changing the Pro+SPA agent per se, rather the SPA feature inside it and only that.

> We're excited to announce an overhaul of our Pro+SPA agent, built to address numerous pain points voiced by our customers. This update tackles critical limitations of the current agent, paving the way for a smoother, more optimized, and future-proof SPA monitoring experience.

Author: @cwli24 How about this:

```suggestion
For customers using the Pro+SPA agent, we're excited to announce an overhaul of SPA monitoring functionality, meant to address numerous customer pain points. This update tackles critical limitations of the current agent, paving the way for a smoother, more optimized, and future-proof SPA monitoring experience.
```

<Callout variant="important">

>     This release is currently in optional trial mode for any browser customer who wants to explore the new features and provide your feedback. The current SPA experience will be gradually phased out as we prepare for a full rollout to all SPA monitoring customers.

Reviewer: Something like restricted or limited preview atm; customers cannot opt in at this time. We decided to curate this list until further notice.

Author: If customers cannot opt in, how do they know if it's changing for them? Is just randomized for various customers?

Reviewer: For APM customers, we work with select ones to enable their appl to get the flag needed to turn things on. For copy & paste customers, they make that config change themselves--no guardrails around this, but we know which customer is supported and who isn't.

* Patchwork solutions: Addressing one issue with SPA patches often introduced another, creating a frustrating cycle of fixes and regressions.
* Third-party library conflicts: Global wrapping, particularly around Promises, often disrupted code functionality due to conflicts with other libraries.
* Performance bottlenecks: Conflicts with code using timers, RAF, and promise chaining led to performance issues, ranging from slowdowns to freezes.

The Pro+SPAv2 agent is designed to eliminate these problems and provide a significantly improved monitoring experience. Key changes include:

* Unwrapped execution: By not wrapping core globals, Pro+SPAv2 unleashes execution speed boosts for your SPAs.
* Aligned with SoftNav heuristics: The new agent adopts Google Chrome's soft navigation heuristics, providing more accurate interaction tracking and improved alignment with browser behavior.
* Simplified interaction determination: Interactions are now defined as UI event (click, keydown, submit) -> route change -> DOM modification, offering a clearer and more efficient capture approach.

> Focus on key metrics: While reported data remains largely unchanged, the new agent no longer tracks JS execution or callback duration within interactions, streamlining the reported information.

Reviewer: Did you mean to make this a bullet point as well?

## Impact and user experience

Changes to the SPA monitoring include:

> * Reduced interaction durations: Expect significantly shorter interaction durations, particularly for route changes. Initial page load interactions should remain stable.

Reviewer: Initial page loads will see a slight reduction

* Potential event disassociation: You might observe previously associated `AjaxRequest` and `JavascriptError` events becoming disassociated from interactions, reflecting the focus on UI-driven interactions.
* API compatibility: The current SPA API remains entirely compatible, ensuring seamless integration with your existing code.
* Enhanced control: The new [`newrelic.interaction({ waitForEnd: true })`](/docs/browser/new-relic-browser/browser-apis/interaction/) API function empowers you to customize interaction completion behavior if needed.

> * Deprecated `createTracer`: While still functional, the `createTracer` function is deprecated as it no longer keeps interactions open or times callbacks. Our data indicates minimal usage of this function.

Reviewer: Should this go under API compatibility as a sub-bullet?

Reviewer: They should note `BrowserTiming` events are also no longer created when using `createTracer` with soft nav

## Try out browser SPA monitoring 2.0 

<Steps>

    <Step>
        ### Review the requirements

        This feature is tested and supported inline with our standard browser support statement <link to it here>.

    </Step>

    <Step>
        ### Add the feature flag to your browser agent

        If your agent was installed with the APM method, contact our support team and we'll enable the feature flag on your account.
        
        If your agent was installed with NPM or the UI copy/paste method, you'll need to add a feature flag in the browser agent code:

            1. Find the New Relic browser code in your webpage HTMl.

            2. In the `init` configuration object, add the `soft_nav` feature flag. Here's an example:

                ```js
                <script type="text/javascript"> ;window.NREUM||(NREUM={});init={ …, feature_flags: ['soft_nav'] }:
                ```

            3. Deploy your app.

        Need to disable this feature? Simply remove the feature flag.

>             

Reviewer: We don't want to directly publish any way for them to manually enable or mess with their configurations atm. In fact, opt-in is restricted atm. I think this section has to change since this page for now will serve as info for the customers who do get it rather than also recruiting existing customers.

    </Step>

    <Step>
        ### Confirm data is being sent to New Relic

        The first thing to check is asking yourself “does my interaction follow the heuristic described above?” Make sure the conditions are met for the default interaction to finish.

        Next, check your network tab in your browser’s dev tools. Filter for requests going to /events/1/. There should be ones for which the payload begins with bel.7;1, indicating your interaction(s) is(are) being captured and sent.

        Finally, interaction data should continue to flow into your browser entity dashboards and charts.

    </Step>

</Steps>

> ## Known issues [#known issues]

Author: ```suggestion
## Known issues [#known-issues]
```

* API usage:

    * `newrelic.interaction().end()` workaround: If you previously used this to address interaction closure issues, you might now see extra, unnecessary interactions. Review your usage to ensure optimal results.
    * `createTracer()` functionality change: `createTracer()` no longer keeps interactions open or times callbacks. If you relied on this for tracking JS code, explore alternative approaches.

* AjaxRequest association:
    * Potential exclusion from interactions: Ajax requests that initiated near the end of previous interactions might now be excluded due to shortened interaction durations.

>     * Manual extension: If you need a request to be attributed to a specific interaction, use the `waitForEnd: true` option within `newrelic.interaction()` and manually end it after the request starts.

Reviewer: maybe re-worded to specify "to keep the interaction open until the request fires, then `.end()` the interaction. This will associate the request to the interaction." if that's more clear

## Leave feedback [#feedback]

For actionable feedback, submit a GitHub issue report any bugs, feature requests, or performance improvements related to SoftNav: [https://github.com/newrelic/newrelic-browser-agent/issues](https://github.com/newrelic/newrelic-browser-agent/issues)

> For general feedback, share your thoughts and suggestions about SoftNav by emailing `browser-agent@newrelic.com` with a subject line starting with "[SoftNav]."

Reviewer: We decided to just lump all feedback into the github issues instead of exposing our team email

freshnessValidatedDate: 2024-03-18
---

<Callout variant="important">
    This release is in limited preview and only available to select customers. Going forward, we plan on gradually phasing out the existing browser SPA experience.
    </Callout> 

For customers monitoring browser single page applications (SPA), we're excited to announce an overhaul of our SPA monitoring functionality, meant to address numerous customer pain points:

* Unusable latest versions: Frequent conflicts with third-party libraries and unreliable interaction capture plagued the existing agent, often rendering the latest version unusable.
* Patchwork solutions: Addressing one issue with SPA patches often introduced another, creating a frustrating cycle of fixes and regressions.
* Third-party library conflicts: Global wrapping, particularly around `Promises`, often disrupted code functionality due to conflicts with other libraries.
* Performance bottlenecks: Conflicts with code using timers, RAF, and promise chaining led to performance issues, ranging from slowdowns to freezes.

> The Pro+SPAv2 agent is designed to eliminate these problems and provide a significantly improved monitoring experience. Key changes include:

Author: ```suggestion
The upgraded SPA monitoring experience is designed to eliminate these problems and provide a significantly improved monitoring experience. Key changes include:
```

Reviewer: "redesigned" or "simplified" may be a more accurate but still positive term

> * Unwrapped execution: By not wrapping core globals, Pro+SPAv2 unleashes execution speed boosts for your SPAs.

Author: ```suggestion
* Unwrapped execution: By not wrapping core globals, the upgrade Pro+SPA agent unleashes execution speed boosts for your SPAs.
```

> * Aligned with soft navigation heuristics: The new agent adopts Google Chrome's soft navigation, providing more accurate interaction tracking and improved alignment with browser behavior.

Author: ```suggestion
* Aligned with soft navigation heuristics: The new experience adopts Google Chrome's soft navigation, providing more accurate interaction tracking and improved alignment with browser behavior.
```

* Simplified interaction determination: Interactions are now defined as a UI event (click/keydown/submit -> route change -> DOM modification), offering a clearer and more efficient capture approach.

> * Focus on key metrics: While reported data remains largely unchanged, the new agent no longer tracks JavaScript execution or callback duration within interactions, streamlining the reported information.

Author: ```suggestion
* Focus on key metrics: While reported data remains largely unchanged, the new experience no longer tracks JavaScript execution or callback duration within interactions, streamlining the reported information.
```

* Reduced interaction durations: Expect significantly shorter interaction durations, particularly for route changes. Initial page loads will see a slight reduction.
* Potential event disassociation: You might observe previously associated `AjaxRequest` and `JavascriptError` events becoming disassociated from interactions, reflecting the focus on UI-driven interactions.

> * New API: The function [`newrelic.interaction({ waitForEnd: true })`](https://docs.newrelic.com/docs/browser/new-relic-browser/browser-apis/interaction/) allows you to customize when a user interaction ends.

Author: @cwli24 I think I added this 
```suggestion
* New API: The function [`newrelic.interaction({ waitForEnd: true })`](new docs link) allows you to customize when a user interaction ends.
```

Reviewer: This is an existing function that's being extended for users with soft nav enabled. This sentence has enough explanation to inform them how to put in the flag and what it does along with our existing doc

> * New API: The function [`newrelic.interaction({ waitForEnd: true })`](https://docs.newrelic.com/docs/browser/new-relic-browser/browser-apis/interaction/) allows you to customize when a user interaction ends.

Author: ```suggestion
* New API: The function `newrelic.interaction({ waitForEnd: true })`allows you to customize when a user interaction ends.
```

* Deprecated API: The function [`createTracer`](https://docs.newrelic.com/docs/browser/new-relic-browser/browser-apis/createtracer/): While still functional, the `createTracer` function is deprecated as it no longer keeps interactions open or times callbacks. Note: `BrowserTiming` events are no longer created when using `createTracer` with the new SPA feature.

## Try out the new browser SPA monitoring experience [#enable-feature]

<Steps>

    <Step>
        ### Review the requirements [#review-requirements]

        This feature is tested and supported in line with our standard [browser support statement](https://docs.newrelic.com/docs/release-notes/new-relic-browser-release-notes/browser-agent-release-notes/#support-statement).

    </Step>

    <Step>
        ### Add the feature flag [#add-feature-flag]

>         If your agent was installed with the [APM method](https://docs.newrelic.com/docs/browser/browser-monitoring/installation/install-browser-monitoring-agent/#apm-agent), contact our support team and we'll enable the feature flag on your account.

Author: ```suggestion
        If your agent was installed with the APM method, contact our support team and we'll enable the feature flag on your account.
```

>         If your agent was installed with NPM or the UI-based, [copy/paste method](https://docs.newrelic.com/docs/browser/browser-monitoring/installation/install-browser-monitoring-agent/#copy-paste), add the following feature flag your browser agent code:

Author: ```suggestion
        If your agent was installed with NPM or the UI-based, copy/paste method, add the following feature flag your browser agent code:
```

            1. Find the New Relic browser agent code in your webpage HTMl.

            2. In the `init` configuration object, add the `soft_nav` feature flag. Here's an example:

                ```js
                <script type="text/javascript"> ;window.NREUM||(NREUM={});init={ …, feature_flags: ['soft_nav'] }:
                ```

            3. Deploy your app.

        Need to disable this feature? Simply remove the feature flag.
            
    </Step>

    <Step>
        ### Confirm data is being sent to New Relic

>         The first thing to check is asking yourself, “Does my interaction follow the heuristic described above?” Make sure the conditions are met for the default interaction to finish.

Author: @cwli24  How do they check that their interaction follows the heuristic/what conditions do they need to meet? It says it's described above, but I'm not sure where.

Reviewer: Stated here:

> * Simplified interaction determination: Interactions are now defined as a UI event (click/keydown/submit -> route change -> DOM modification), offering a clearer and more efficient capture approach.

>         The first thing to check is asking yourself, “Does my interaction follow the heuristic described above?” Make sure the conditions are met for the default interaction to finish.

Author: ```suggestion
        First, check to make sure your interactions follow the soft navigation heuristic (click/keydown/submit -> route change -> DOM modification).
```
